### PR TITLE
Merge currently staged work into master

### PR DESCRIPTION
**Pull request by Tom-Goring**
A lot in these commits, the merge was very awkward so hopefully the commits are roughly representative of the changes.

I'll try to list the changes here:
- Role based routing is here, along with the base JSPs for the different roles. Login and Register can only be accessed if not logged in, else the user will be redirected to /dashboard, which will resolve into the JSP appropriate to their type.
- This means the user class and table has been expanded to have a role field.
- First iteration of audit logging is here - add a log to the audit using `Log.info("message describing what happened")`. The audit log is printed in the admin JSP by using a custom dispatcher in the DashboardFilter.
- The admin page now has a button to recreate the database tables - which will be removed when we're finished. Currently an admin user is added on table recreation for ease of access - I will likely add a config file entry for determining the password of that user.
- The backend models and table for appointments is here - but currently they are unused. This will need to be added in a future PR. The appointments API may still change depending on ease of testing and integration into views - Max is investigating.
- The ApplicationDAO has been changed into using only static methods in order to decouple the logic a bit more.
- Downgraded our web.xml version to 3.0 in order that very old versions of netbeans can function correctly.
- Session duration has been changed to 5 minutes

If you have questions about how things work or why things are the way they are - do say. Especially the logic I've used in LoginFilter feels very janky, so another set of eyes would be very welcome. It would be good if everybody could run it on their machine to ensure it all works - you will have to log in as username: admin password: admin and click recreate tables to get the full functionality.

### CODE REVIEW

File: src/main/webapp/dashboards/patient.jsp
Comment (CharlievWilliams) [inline src/main/webapp/dashboards/patient.jsp:11]: If the CSS PR gets merged in first we'll need to add the updates to these new JSP headers.

File: src/main/java/Models/Appointment/Appointment.java
Comment (CharlievWilliams) [inline src/main/java/Models/Appointment/Appointment.java:75]: There are a bunch of formatting / naming comments I could make in this PR, but I think in the interest of time I'll refrain for now and pick it up in a quick PR at the end of this sprint or next sprint

Comment (Tom-Goring) [reply src/main/java/Models/Appointment/Appointment.java:75]: Yeah, we'll run a cleanup PR after implementations are done I imagine - I fixed most of the linting issues in Max's code but didn't touch most variable names etc

File: src/main/java/Models/Appointment/Appointment.java
Comment (CharlievWilliams) [inline src/main/java/Models/Appointment/Appointment.java:47]: This likely isn't a requirement but might be something to be aware of.

At my previous job, we ran into issues with this method (Or similar) when working with multiple time zones. For example, if a patient were to book a holiday online while abroad. This function (Or similar) would return the local time and make the whole journey go a bit spoopy.

Maybe we could talk about this when going for those 80+ marks in the final sprint.

File: src/main/java/Filters/DashboardFilter.java
Comment (CharlievWilliams) [inline src/main/java/Filters/DashboardFilter.java:50]: Could you talk me through the logic of this if statement, please? Why is it that we would need to recreate tables when redirecting to admin dashboard?

File: src/main/java/Filters/LoginFilter.java
Comment (CharlievWilliams) [inline src/main/java/Filters/LoginFilter.java:27]: Could you talk through this logic a bit too, please?

File: src/main/java/Models/User/User.java
Comment (CharlievWilliams) [inline src/main/java/Models/User/User.java:24]: Do we actually plan to include the above commented values?

File: src/main/java/Utils/Tables.java
Comment (CharlievWilliams) [inline src/main/java/Utils/Tables.java:22]: Are we using `firstName` as the username people log in with? Perhaps we should make that clearer in a future PR, as I wouldn't normally say those terms are interchangeable?

File: src/main/webapp/WEB-INF/web.xml
Comment (CharlievWilliams) [inline src/main/webapp/WEB-INF/web.xml:14]: Will the addition of the wildcard `*` work in place of the `/index` we introduced in the CSS PR?

File: src/main/webapp/WEB-INF/web.xml
Comment (CharlievWilliams) [inline src/main/webapp/WEB-INF/web.xml:23]: Why do we need a `*` here but not on the `RegistrationServlet` URL pattern?

File: src/main/java/Filters/DashboardFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/DashboardFilter.java:50]: I presume you are familiar with how browsers send HTTP methods? If not - when you navigate to a URL, the browser sends a GET request which usually returns the page (though you can be evil and do other stuff). A normal website with javascript would use something like Ajax to send a POST request to a URL endpoint - so a POST request to something like example.com/api/recreateTables. Here, I didn't want a servlet specifically sitting on a route purely dedicated to recreating tables, so instead, if a post request is sent to the admin page we use that as a sign to trigger the recreation of the tables. If we wanted multiple POST functionalities sitting there, we could add the specific functionality we want to trigger as a parameter or on the body of the POST request and expand the if to include a match on that data.

File: src/main/java/Filters/LoginFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/LoginFilter.java:27]: Most websites will redirect a logged in user to their dashboard page, for example, Facebook will redirect you to your feed if you try to access its login page while you have a session. In order to replicate this, we check if the user is logged in, if they are, we check if they are trying to access /login or /register and redirect them to their dashboard if they are. If they are logged in and trying to access something else, we forward the request and give them access to whatever they are trying to get into as usual. If they are sending a POST request to /login, this is treated as them wishing to log OUT and thus allows them to perform the request as usual, resulting in them being logged out due to the code on the /login servlet.

File: src/main/java/Models/User/User.java
Comment (Tom-Goring) [reply src/main/java/Models/User/User.java:24]: Currently I am unsure - they are not specifically required, so can be removed in the repo if we like.

File: src/main/java/Utils/Tables.java
Comment (Tom-Goring) [reply src/main/java/Utils/Tables.java:22]: The user logs in with a combination of their email and password (though the PK for a user is still an auto-incremented integer). Emails are required to be unique and an error will be given if somebody tries to register with a duplicate email.

File: src/main/webapp/WEB-INF/web.xml
Comment (Tom-Goring) [reply src/main/webapp/WEB-INF/web.xml:14]: The /* means the LoginFilter will work on every page, as we need to know where to route somebody when they request any page, not just privileged ones - this is to enable us to redirect /login and /register requests to their dashboard for ergonomics

File: src/main/webapp/WEB-INF/web.xml
Comment (Tom-Goring) [reply src/main/webapp/WEB-INF/web.xml:23]: Hmm, not really sure to be honest - it should really be set up to redirect all /dashboard/* stuff to just /dashboard, though it currently just gives you the right page but for weird urls like /dashboard/weirdurls.
image

File: src/main/java/Models/Appointment/Appointment.java
Comment (Tom-Goring) [reply src/main/java/Models/Appointment/Appointment.java:47]: Dealing with timezones is always a nightmare lol

File: src/main/java/Filters/DashboardFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/DashboardFilter.java:50]: It just looks and feels weird because we can only use forms to make POST requests, even when we don't want to load into a different page, just launch some desired functionality. The page reloading is why websites don't really do this anymore, as it leads to weird stuff like pressing a button, the whole page reloading + the functionality of the button happening, then pressing refresh and it 'pressing' the button again.

File: src/main/java/Filters/LoginFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/LoginFilter.java:27]: Here is a flowchart of my intended logic, hopefully this makes it clearer - the code is pretty nasty tbh
image

Comment (Tom-Goring) [reply src/main/java/Filters/LoginFilter.java:27]: This does miss the non logged in user trying to access /register <.<

Comment (Tom-Goring) [reply src/main/java/Filters/LoginFilter.java:27]: Okay, flow chart 2 electric boogaloo:
image
Also in the making of this realised I missed an if check for the registration page while logged in, so thanks for asking!

File: src/main/webapp/WEB-INF/web.xml
Comment (Tom-Goring) [reply src/main/webapp/WEB-INF/web.xml:14]: See the above flowchart for the logic of what goes on behind the scenes everytime somebody tries to access any page.

File: src/main/java/Utils/Tables.java
Comment (CharlievWilliams) [reply src/main/java/Utils/Tables.java:22]: I see! I got confused because the admin user that you created and mention at the top of the conversation has a username: admin which isn't an email.

File: src/main/java/Models/User/User.java
Comment (CharlievWilliams) [reply src/main/java/Models/User/User.java:24]: Perhaps we can discuss this in our next standup.

File: src/main/java/Utils/Tables.java
Comment (Tom-Goring) [reply src/main/java/Utils/Tables.java:22]: Yeah in the future the initial admin account will be configurable in the config file

File: src/main/java/Filters/DashboardFilter.java
Comment (CharlievWilliams) [reply src/main/java/Filters/DashboardFilter.java:55]: Thanks for explaining mate. I'm guessing the parameter method is technically the "correct" way to go about this, even for a single functionality? Either way, I'm happy to continue with what we have here for time's sake!

File: src/main/java/Filters/LoginFilter.java
Comment (CharlievWilliams) [reply src/main/java/Filters/LoginFilter.java:27]: Similarly here, I suppose a named parameter in the POST would make this clearer? Instead of making those changes, maybe adding some comments in the code for this section would certainly help me understand.

**Comment by CharlievWilliams** [comment]
@Tom-Goring I'm happy to approve this if you add some comments to the cursed code talking through the process for when we come across it later.

- [ ] Add comments
- [ ] Create cleanup ticket
- [ ] Add CSS to JSP (Or ignore if merging first)

File: src/main/java/Filters/LoginFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/LoginFilter.java:27]: I'll think about adding a comment to the area when I'm done with the calendar stuff

File: src/main/java/Filters/DashboardFilter.java
Comment (Tom-Goring) [reply src/main/java/Filters/DashboardFilter.java:55]: It would be more obvious what was going on if it had an entire servlet behind it, but we'd end up with a lot of servlets that do very little. In the future if it was expanded it would likely be moved into a Servlet and a POST would match parameters for functionality

**Comment by Tom-Goring** [comment]
Rebased onto master, comments added - I will merge when I get 2 approvals.

**Review by CharlievWilliams — APPROVED** [review]

**Review by Maxwell2Gyamfi — APPROVED** [review]
All looks good to me
Only one question on the update appointment, would it be possible to update , date and time too?

**Pull request merged**
